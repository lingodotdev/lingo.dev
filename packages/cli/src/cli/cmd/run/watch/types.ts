import { CmdRunContext } from "../_types";
import type { Stats } from "fs";

// Core watch configuration interface
export interface WatchConfiguration {
  patterns: {
    include: string[];
    exclude: string[];
  };
  debounce: {
    delay: number;
    maxWait?: number;
  };
  monitoring: {
    enableProgressIndicators: boolean;
    enableNotifications: boolean;
    logLevel: "silent" | "minimal" | "verbose";
  };
  performance: {
    batchSize: number;
    rateLimitDelay: number;
  };
}

// Watch status and state interfaces
export interface WatchStatus {
  isActive: boolean;
  watchedFiles: string[];
  pendingChanges: string[];
  lastActivity: Date;
  errorCount: number;
}

export interface WatchState {
  isRunning: boolean;
  isRetranslating: boolean;
  pendingChanges: Map<string, FileChangeEvent>;
  watchedPatterns: string[];
  startTime: Date;
  lastRetranslation?: Date;
  statistics: WatchStatistics;
}

export interface WatchStatistics {
  totalChanges: number;
  retranslationCount: number;
  errorCount: number;
  averageRetranslationTime: number;
}

// Enhanced watch context
export interface EnhancedWatchContext extends CmdRunContext {
  watchConfig: WatchConfiguration;
  watchState: WatchState;
  feedbackManager: FeedbackManager;
}

// File change event interface
export interface FileChangeEvent {
  type: "add" | "change" | "unlink" | "addDir" | "unlinkDir";
  path: string;
  timestamp: Date;
  stats?: Stats;
}

// Watch options for file watcher service
export interface WatchOptions {
  debounceDelay: number;
  ignoreInitial: boolean;
  awaitWriteFinish: {
    stabilityThreshold: number;
    pollInterval: number;
  };
  ignored: string[];
}

// Debouncing strategy configuration
export interface DebounceStrategy {
  type: "simple" | "adaptive" | "batch";
  delay: number;
  maxWait?: number;
  batchSize?: number;
  rateLimitDelay?: number;
}

// Translation progress and result interfaces
export interface TranslationProgress {
  totalTasks: number;
  completedTasks: number;
  currentTask?: string;
  estimatedTimeRemaining?: number;
}

export interface TranslationResult {
  success: boolean;
  duration: number;
  tasksCompleted: number;
  errors: WatchError[];
}

// Error handling interfaces
export interface WatchError {
  type: "file_system" | "translation" | "configuration";
  message: string;
  path?: string;
  recoverable: boolean;
}

export interface ValidationResult {
  isValid: boolean;
  errors: string[];
  warnings: string[];
}

// Pattern resolution interfaces
export interface ResolvedPatterns {
  include: string[];
  exclude: string[];
  resolved: string[];
}

// Error recovery strategy interface
export interface ErrorRecoveryStrategy {
  maxRetries: number;
  backoffMultiplier: number;
  recoveryActions: RecoveryAction[];
}

export interface RecoveryAction {
  condition: (error: Error) => boolean;
  action: (error: Error) => Promise<void>;
  description: string;
}

// Forward declarations for interfaces (will be implemented in respective files)
export interface WatchManager {
  start(ctx: CmdRunContext): Promise<void>;
  stop(): Promise<void>;
  getStatus(): WatchStatus;
  updateConfiguration(config: WatchConfiguration): Promise<void>;
}

export interface ConfigurationManager {
  loadConfiguration(ctx: CmdRunContext): Promise<WatchConfiguration>;
  validateConfiguration(config: WatchConfiguration): ValidationResult;
  mergeWithDefaults(config: Partial<WatchConfiguration>): WatchConfiguration;
}

export interface FileWatcherService {
  initialize(patterns: string[], options: WatchOptions): Promise<void>;
  addPattern(pattern: string): Promise<void>;
  removePattern(pattern: string): Promise<void>;
  getWatchedFiles(): string[];
  destroy(): Promise<void>;
}

export interface DebounceController {
  scheduleRetranslation(changes: FileChangeEvent[]): void;
  cancelPending(): void;
  isRetranslationPending(): boolean;
  getTimeUntilExecution(): number;
}

export interface FeedbackManager {
  reportFileChange(event: FileChangeEvent): void;
  reportRetranslationStart(files: string[]): void;
  reportRetranslationProgress(progress: TranslationProgress): void;
  reportRetranslationComplete(result: TranslationResult): void;
  reportError(error: WatchError): void;
}

export interface PatternResolver {
  resolveWatchPatterns(ctx: CmdRunContext): Promise<ResolvedPatterns>;
  validatePatterns(patterns: string[]): ValidationResult;
  expandGlobPatterns(patterns: string[]): Promise<string[]>;
}
