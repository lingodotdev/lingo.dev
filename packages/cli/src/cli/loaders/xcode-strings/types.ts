export enum TokenType {
  COMMENT_SINGLE = "COMMENT_SINGLE",
  COMMENT_MULTI = "COMMENT_MULTI",
  STRING = "STRING",
  EQUALS = "EQUALS",
  SEMICOLON = "SEMICOLON",
  EOF = "EOF",
}

export interface Token {
  type: TokenType;
  value: string;
  line: number;
  column: number;
}

export interface Position {
  line: number;
  column: number;
}
