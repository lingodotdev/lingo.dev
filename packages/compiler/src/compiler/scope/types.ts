import * as t from '@babel/types';
import { NodePath } from '@babel/core';
import { ReplexicaScopeData, ReplexicaScopeHint } from '../types';

export interface IReplexicaScope {
  get id(): string;
  injectIntl(fileId: string, isServer: boolean, supportedLocales: string[]): ReplexicaScopeData;
  extractHints(): ReplexicaScopeHint[];
}

export type ReplexicaScopeExtractor = {
  fromNode(path: NodePath<t.Node>): IReplexicaScope[];
}