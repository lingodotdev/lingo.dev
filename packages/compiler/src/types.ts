import { ReplexicaScopeHint } from "./compiler";

export type ReplexicaLocaleData = {
  [fileId: string]: {
    [scopeId: string]: {
      [chunkId: string]: string;
    };
  };
};

export type ReplexicaData = {
  settings: ReplexicaSettingsData;
  meta: ReplexicaMetaData;
};

export type ReplexicaMetaData = {
  files: ReplexicaMetaFileData;
  scopes: ReplexicaMetaScopeData;
};

export type ReplexicaSettingsData = {
  locale: {
    source: string;
    targets: string[];
  };
};

export type ReplexicaMetaFileData = {
  [fileId: string]: {
    isClient: boolean;
  };
};

export type ReplexicaMetaScopeData = {
  [scopeId: string]: {
    hints: ReplexicaScopeHint[];
  };
};