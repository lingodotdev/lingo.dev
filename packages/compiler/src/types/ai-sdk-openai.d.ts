declare module "@ai-sdk/openai" {
  export function createOpenAI(options?: any): any;
}
