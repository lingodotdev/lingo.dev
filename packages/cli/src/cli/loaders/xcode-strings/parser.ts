import { Token, TokenType } from "./types";
import { unescapeString } from "./escape";

export class Parser {
  private tokens: Token[];
  private pos: number;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
    this.pos = 0;
  }

  parse(): Record<string, string> {
    const result: Record<string, string> = {};

    while (this.pos < this.tokens.length) {
      const token = this.current();

      // Skip comments
      if (
        token.type === TokenType.COMMENT_SINGLE ||
        token.type === TokenType.COMMENT_MULTI
      ) {
        this.advance();
        continue;
      }

      // End of file
      if (token.type === TokenType.EOF) {
        break;
      }

      // Expect entry: STRING "=" STRING ";"
      if (token.type === TokenType.STRING) {
        const entry = this.parseEntry();
        if (entry) {
          result[entry.key] = entry.value;
        }
        continue;
      }

      // Skip unexpected tokens gracefully
      this.advance();
    }

    return result;
  }

  private parseEntry(): { key: string; value: string } | null {
    // Current token should be STRING (key)
    const keyToken = this.current();
    if (keyToken.type !== TokenType.STRING) {
      return null;
    }
    const key = keyToken.value;
    this.advance();

    // Expect '='
    if (!this.expect(TokenType.EQUALS)) {
      // Missing '=' - skip this entry
      return null;
    }

    // Expect STRING (value)
    const valueToken = this.current();
    if (valueToken.type !== TokenType.STRING) {
      // Missing value - skip this entry
      return null;
    }
    const rawValue = valueToken.value;
    this.advance();

    // Expect ';'
    if (!this.expect(TokenType.SEMICOLON)) {
      // Missing ';' - but still process the entry
      // (more forgiving)
    }

    // Unescape the value
    const value = unescapeString(rawValue);

    return { key, value };
  }

  private current(): Token {
    return this.tokens[this.pos];
  }

  private advance(): void {
    if (this.pos < this.tokens.length) {
      this.pos++;
    }
  }

  private expect(type: TokenType): boolean {
    if (this.current()?.type === type) {
      this.advance();
      return true;
    }
    return false;
  }
}
