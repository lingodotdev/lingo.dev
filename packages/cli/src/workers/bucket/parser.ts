import { parseStringPromise, Builder } from 'xml2js';
import YAML from 'yaml';
import _ from 'lodash';
import GrayMatter from 'gray-matter';
import { flatten, unflatten } from 'flat';

export interface IBucketParser {
  deserialize: (locale: string, content: any) => Promise<Record<string, string>>;
  serialize: (locale: string, content: Record<string, string | number | boolean>) => Promise<any>;
}

export function createJsonParser(): IBucketParser {
  return {
    async deserialize(locale: string, content: string) {
      const parsedJson = JSON.parse(content);
      const flattenedJson: Record<string, string> = flatten(parsedJson, {
        delimiter: '/',
        transformKey(key) {
          // apply uri encoding to keys
          return encodeURIComponent(String(key));
        },
      });

      return flattenedJson;
    },
    async serialize(locale: string, payload: Record<string, string | number | boolean>) {
      const unflattenedPayload = unflatten(payload, {
        delimiter: '/',
        transformKey(key) {
          // apply uri decoding to keys
          return decodeURIComponent(String(key));
        }
      });

      const content = JSON.stringify(unflattenedPayload, null, 2);
      return content;
    }
  };
}

export function createMarkdownParser(): IBucketParser {
  // Define markdown section starter patterns
  const sectionStarters = [
    // Matches headings
    /^#+\s.*$/gm, 
    // Matches dash dividers
    /^[-]{3,}$/gm, 
    // Matches equals dividers
    /^[=]{3,}$/gm, 
    // Matches asterisk dividers
    /^[*]{3,}$/gm, 
    // Matches images that take up a line
    /^!\[.*\]\(.*\)$/gm,
    // Matches links that take up a line
    /^\[.*\]\(.*\)$/gm,
  ];

  return {
    async deserialize(locale: string, rawContent: string) {
      const fmContent = GrayMatter(rawContent);

      const attributes = fmContent.data;
      const content = fmContent.content;

      // Combine all patterns into a single regex
      const combinedPattern = new RegExp(sectionStarters.map(pattern => `(${pattern.source})`).join('|'), 'gm');

      // Split content into sections based on the combined regex
      const sections = content.split(combinedPattern).filter(Boolean);

      // Group sections into a record with section index as key
      const bodyPartial: Record<string, string> = sections.reduce((acc, section, index) => ({
        ...acc,
        [`markdown-line-${index}`]: section,
      }), {} as any);

      // Attributes partial, if any, prefixed with frontmatter-attribute-[key]
      const attributesPartial = Object.entries(attributes).reduce((acc, [key, value]) => ({
        ...acc,
        [`frontmatter-attribute-${key}`]: value,
      }), {} as any);

      const result = {
        ...bodyPartial,
        ...attributesPartial,
      };

      return result;
    },
    async serialize(locale: string, content: Record<string, string | number | boolean>) {
      const body = Object
        .entries(content)
        .filter(([key]) => key.startsWith('markdown-line-'))
        .map(([, value]) => value)
        .join('');

      const attributes = Object
        .entries(content)
        .filter(([key]) => key.startsWith('frontmatter-attribute-'))
        .reduce((acc, [key, value]) => ({
          ...acc,
          [key.replace('frontmatter-attribute-', '')]: value,
        }), {} as any);

      const result = GrayMatter.stringify(body, attributes);

      return result;
    },
  };
}

export type YamlBucketParserParams = {
  rootKey: boolean;
};
export function createYamlParser(params: YamlBucketParserParams): IBucketParser {
  return {
    async deserialize(locale: string, content: any) {
      const parsed = YAML.parse(content);
      return params.rootKey ? parsed[locale] : parsed;
    },
    async serialize(locale: string, content: Record<string, string | number | boolean>) {
      return YAML.stringify(params.rootKey ? { [locale]: content } : content);
    }
  };
}

export function createXcodeParser() {
  let _existingData: any;
  return {
    async deserialize(locale: string, content: any) {
      _existingData = await JSON.parse(content);

      const resultData: Record<string, any> = {};

      for (const [translationKey, _translationEntity] of Object.entries(_existingData.strings)) {
        const rootTranslationEntity = _translationEntity as any;
        const langTranslationEntity = rootTranslationEntity?.localizations?.[locale];
        if (langTranslationEntity) {
          if ('stringUnit' in langTranslationEntity) {
            resultData[translationKey] = langTranslationEntity.stringUnit.value;
          } else if ('variations' in langTranslationEntity) {
            if ('plural' in langTranslationEntity.variations) {
              resultData[translationKey] = {
                one: langTranslationEntity.variations.plural.one?.stringUnit?.value || '',
                other: langTranslationEntity.variations.plural.other?.stringUnit?.value || '',
                zero: langTranslationEntity.variations.plural.zero?.stringUnit?.value || '',
              };
            }
          }
        }
      }

      return resultData;
    },
    async serialize(locale: string, payload: Record<string, any>): Promise<string> {
      const langDataToMerge: any = {};
      langDataToMerge.strings = {};

      for (const [key, value] of Object.entries(payload)) {
        if (typeof value === 'string') {
          langDataToMerge.strings[key] = {
            extractionState: 'manual',
            localizations: {
              [locale]: {
                stringUnit: {
                  state: 'translated',
                  value,
                },
              },
            },
          };
        } else {
          langDataToMerge.strings[key] = {
            extractionState: 'manual',
            localizations: {
              [locale]: {
                variations: {
                  plural: {
                    one: {
                      stringUnit: {
                        state: 'translated',
                        value: value.one,
                      },
                    },
                    other: {
                      stringUnit: {
                        state: 'translated',
                        value: value.other,
                      },
                    },
                    zero: {
                      stringUnit: {
                        state: 'translated',
                        value: value.zero,
                      },
                    },
                  },
                },
              },
            },
          };
        }
      }

      const resultData = _.merge({}, _existingData, langDataToMerge);
      return JSON.stringify(resultData, null, 2);
    }
  };
}

export function createAndroidParser(): IBucketParser {
  return {
    async deserialize(locale: string, content: any) {
      const parsedResult = await parseStringPromise(content);
      const resources = parsedResult.resources;

      const result: Record<string, any> = {};

      // Parse single strings
      if (resources.string) {
        resources.string.forEach((item: any) => {
          result[item.$.name] = item._ || '';
        });
      }

      // Parse string arrays
      if (resources['string-array']) {
        resources['string-array'].forEach((item: any) => {
          result[item.$.name] = item.item.map((i: any) => i._ || '');
        });
      }

      // Parse plurals
      if (resources.plurals) {
        resources.plurals.forEach((item: any) => {
          const plurals: Record<string, string> = {};
          item.item.forEach((i: any) => {
            plurals[i.$.quantity] = i._ || '';
          });
          result[item.$.name] = plurals;
        });
      }

      return result;
    },

    async serialize(locale: string, content: Record<string, any>) {
      const builder = new Builder({ headless: true })
      const resources: any = {
        resources: {
          string: [],
          'string-array': [],
          plurals: [],
        },
      };

      for (const key in content) {
        if (typeof content[key] === 'string') {
          resources.resources.string.push({ $: { name: key }, _: content[key] });
        } else if (Array.isArray(content[key])) {
          resources.resources['string-array'].push({
            $: { name: key },
            item: content[key].map((value: string) => ({ _: value })),
          });
        } else if (typeof content[key] === 'object') {
          const pluralItems = [];
          for (const quantity in content[key]) {
            pluralItems.push({
              $: { quantity },
              _: content[key][quantity],
            });
          }
          resources.resources.plurals.push({
            $: { name: key },
            item: pluralItems,
          });
        }
      }

      // Remove empty sections if they have no items
      if (resources.resources.string.length === 0) {
        delete resources.resources.string;
      }
      if (resources.resources['string-array'].length === 0) {
        delete resources.resources['string-array'];
      }
      if (resources.resources.plurals.length === 0) {
        delete resources.resources.plurals;
      }

      const xml = builder.buildObject(resources);
      return xml;
    },
  };
}