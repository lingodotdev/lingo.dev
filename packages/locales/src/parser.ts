import type { LocaleComponents, LocaleDelimiter, ParseResult } from "./types";
import { LOCALE_REGEX } from "./constants";

/**
 * Normalizes the case of locale components before parsing
 *
 * @param locale - The locale string to normalize
 * @returns The normalized locale string
 *
 * @example
 * normalizeLocaleCase("EN-US")     // "en-US"
 * normalizeLocaleCase("en-us")     // "en-US"
 * normalizeLocaleCase("zh-hans-cn") // "zh-Hans-CN"
 */
function normalizeLocaleCase(locale: string): string {
  // Split by either hyphen or underscore
  const parts = locale.split(/[-_]/);

  if (parts.length === 1) {
    // Language only: normalize to lowercase
    return parts[0].toLowerCase();
  }

  if (parts.length === 2) {
    // Language-region: normalize language to lowercase, region to uppercase
    const language = parts[0].toLowerCase();
    const region = parts[1].toUpperCase();
    return `${language}-${region}`;
  }

  if (parts.length === 3) {
    // Language-script-region: normalize language to lowercase, preserve script case, region to uppercase
    const language = parts[0].toLowerCase();
    const script = parts[1]; // Preserve original case as-is
    const region = parts[2].toUpperCase();
    return `${language}-${script}-${region}`;
  }

  // For any other number of parts, return as-is
  return locale;
}

/**
 * Breaks apart a locale string into its components
 *
 * @param locale - The locale string to parse
 * @returns LocaleComponents object with language, script, and region
 *
 * @example
 * ```typescript
 * parseLocale("en-US");          // { language: "en", region: "US" }
 * parseLocale("en_US");          // { language: "en", region: "US" }
 * parseLocale("zh-Hans-CN");     // { language: "zh", script: "Hans", region: "CN" }
 * parseLocale("zh_Hans_CN");     // { language: "zh", script: "Hans", region: "CN" }
 * parseLocale("es");             // { language: "es" }
 * parseLocale("sr-Cyrl-RS");     // { language: "sr", script: "Cyrl", region: "RS" }
 * ```
 */
export function parseLocale(locale: string): LocaleComponents {
  if (typeof locale !== "string") {
    throw new Error("Locale must be a string");
  }

  if (!locale.trim()) {
    throw new Error("Locale cannot be empty");
  }

  // Normalize case before parsing:
  // - Language: convert to lowercase (e.g., "EN" -> "en")
  // - Script: preserve case (e.g., "Hans", "hans" -> "Hans")
  // - Region: convert to uppercase (e.g., "us" -> "US")
  const normalizedLocale = normalizeLocaleCase(locale);

  const match = normalizedLocale.match(LOCALE_REGEX);

  if (!match) {
    throw new Error(`Invalid locale format: ${locale}`);
  }

  const [, language, script, region] = match;

  const components: LocaleComponents = {
    language: language.toLowerCase(),
  };

  // Add script if present
  if (script) {
    components.script = script;
  }

  // Add region if present
  if (region) {
    components.region = region.toUpperCase();
  }

  return components;
}

/**
 * Parses a locale string and returns detailed information about the parsing result
 *
 * @param locale - The locale string to parse
 * @returns ParseResult with components, delimiter, and validation info
 */
export function parseLocaleWithDetails(locale: string): ParseResult {
  try {
    const components = parseLocale(locale);

    // Determine the delimiter used
    let delimiter: LocaleDelimiter | null = null;
    if (locale.includes("-")) {
      delimiter = "-";
    } else if (locale.includes("_")) {
      delimiter = "_";
    }

    return {
      components,
      delimiter,
      isValid: true,
    };
  } catch (error) {
    return {
      components: { language: "" },
      delimiter: null,
      isValid: false,
      error: error instanceof Error ? error.message : "Unknown parsing error",
    };
  }
}

/**
 * Extracts just the language code from a locale string
 *
 * @param locale - The locale string to parse
 * @returns The language code
 *
 * @example
 * ```typescript
 * getLanguageCode("en-US");      // "en"
 * getLanguageCode("zh-Hans-CN"); // "zh"
 * getLanguageCode("es-MX");      // "es"
 * getLanguageCode("fr_CA");      // "fr"
 * ```
 */
export function getLanguageCode(locale: string): string {
  return parseLocale(locale).language;
}

/**
 * Extracts the script code from a locale string
 *
 * @param locale - The locale string to parse
 * @returns The script code or null if not present
 *
 * @example
 * ```typescript
 * getScriptCode("zh-Hans-CN");   // "Hans"
 * getScriptCode("zh-Hant-TW");   // "Hant"
 * getScriptCode("sr-Cyrl-RS");   // "Cyrl"
 * getScriptCode("en-US");        // null
 * getScriptCode("es");           // null
 * ```
 */
export function getScriptCode(locale: string): string | null {
  const components = parseLocale(locale);
  return components.script || null;
}

/**
 * Extracts the region/country code from a locale string
 *
 * @param locale - The locale string to parse
 * @returns The region code or null if not present
 *
 * @example
 * ```typescript
 * getRegionCode("en-US");        // "US"
 * getRegionCode("zh-Hans-CN");   // "CN"
 * getRegionCode("es");           // null
 * getRegionCode("fr_CA");        // "CA"
 * ```
 */
export function getRegionCode(locale: string): string | null {
  const components = parseLocale(locale);
  return components.region || null;
}
