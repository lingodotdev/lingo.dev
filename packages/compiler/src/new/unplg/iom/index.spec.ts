import { describe, expect, it } from 'vitest';
import { parse } from '@babel/parser';
import * as t from '@babel/types';
import { extractI18n } from '.';

describe('parse-i18n-scope', () => {
  function generateAstFromCode(code: string): t.File {
    return parse(code, { sourceType: 'module', plugins: ['jsx'] });
  }

  it('simple jsx element', () => {
    const code = `<title>My app</title>`;

    const ast = generateAstFromCode(code);
    const scope = extractI18n(ast);
    const scopeObj = scope?.toJSON();

    expect(scopeObj).toEqual({
      role: 'scope',
      type: 'js/program',
      explicit: false,
      hint: '',
      fragments: [],
      scopes: [{
        role: 'scope',
        type: 'jsx/element',
        explicit: false,
        hint: '',
        fragments: [
          { role: 'fragment', type: 'jsx/text', value: 'My app' }
        ],
        scopes: [],
      }],
    });
  });

  it('jsx element with explicit data-i18n attribute', () => {
    const code = `
      <header data-i18n>
        <h1>My app</h1>
        <p>This is a demo app</p>
      </header>
    `;

    const ast = generateAstFromCode(code);
    const scope = extractI18n(ast);
    const scopeObj = scope?.toJSON();

    expect(scopeObj).toEqual({
      role: 'scope',
      type: 'js/program',
      explicit: false,
      hint: '',
      fragments: [],
      scopes: [{
        role: 'scope',
        type: 'jsx/element',
        explicit: true,
        hint: '',
        fragments: [],
        scopes: [
          {
            role: 'scope',
            type: 'jsx/element',
            explicit: false,
            hint: '',
            fragments: [
              { role: 'fragment', type: 'jsx/text', value: 'My app' }
            ],
            scopes: [],
          },
          {
            role: 'scope',
            type: 'jsx/element',
            explicit: false,
            hint: '',
            fragments: [
              { role: 'fragment', type: 'jsx/text', value: 'This is a demo app' }
            ],
            scopes: [],
          }
        ],
      }],
    });
  });

  it('jsx element with text and sub-elements', () => {
    const code = `
      <p>
        Some other text
        <span>© 2030</span>
      </p>
    `;

    const ast = generateAstFromCode(code);
    const scope = extractI18n(ast);
    const scopeObj = scope?.toJSON();

    expect(scopeObj).toEqual({
      role: 'scope',
      type: 'js/program',
      explicit: false,
      hint: '',
      fragments: [],
      scopes: [{
        role: 'scope',
        type: 'jsx/element',
        explicit: false,
        hint: '',
        fragments: [
          { role: 'fragment', type: 'jsx/text', value: 'Some other text' },
          { role: 'fragment', type: 'jsx/text', value: '© 2030' }
        ],
        scopes: [],
      }],
    });
  });

  it('explicit data-18n doesnt trigger a scope if the element would be a scope anyway', () => {
    const code = `
      <p data-i18n>Something</p>
    `;

    const ast = generateAstFromCode(code);
    const scope = extractI18n(ast);
    const scopeObj = scope?.toJSON();

    expect(scopeObj).toEqual({
      role: 'scope',
      type: 'js/program',
      explicit: false,
      hint: '',
      fragments: [],
      scopes: [{
        role: 'scope',
        type: 'jsx/element',
        explicit: false,
        hint: '',
        fragments: [
          { role: 'fragment', type: 'jsx/text', value: 'Something' }
        ],
        scopes: [],
      }],
    });
  });

  it('nested jsx elements', () => {
    const code = `
      <main>
        <p>
          Click
          <a href="/more">
            here
          </a>
          to learn more
        </p>
        <span>Some text</span>
      </main>
    `;

    const ast = generateAstFromCode(code);
    const scope = extractI18n(ast);
    const scopeObj = scope?.toJSON();

    expect(scopeObj).toEqual({
      role: 'scope',
      type: 'js/program',
      explicit: false,
      hint: '',
      fragments: [],
      scopes: [
        {
          role: 'scope',
          type: 'jsx/element',
          explicit: false,
          hint: '',
          fragments: [
            { role: 'fragment', type: 'jsx/text', value: 'Click' },
            { role: 'fragment', type: 'jsx/text', value: 'here' },
            { role: 'fragment', type: 'jsx/text', value: 'to learn more' }
          ],
          scopes: [],
        },
        {
          role: 'scope',
          type: 'jsx/element',
          explicit: false,
          hint: '',
          fragments: [
            { role: 'fragment', type: 'jsx/text', value: 'Some text' }
          ],
          scopes: [],
        }
      ],
    });
  });

  it('jsx element with an commonly localizable attribute', () => {
    const code = `
      <span title="Some label for the span">Some text</span>
    `;

    const ast = generateAstFromCode(code);
    const scope = extractI18n(ast);
    const scopeObj = scope?.toJSON();

    expect(scopeObj).toEqual({
      role: 'scope',
      type: 'js/program',
      explicit: false,
      hint: '',
      fragments: [],
      scopes: [
        {
          role: 'scope',
          type: 'jsx/element',
          explicit: false,
          hint: '',
          fragments: [
            { role: 'fragment', type: 'jsx/text', value: 'Some text' }
          ],
          scopes: [
            {
              role: 'scope',
              type: 'jsx/attribute',
              explicit: false,
              hint: '',
              fragments: [
                { role: 'fragment', type: 'js/text', value: 'Some label for the span' }
              ],
              scopes: [],
            }
          ],
        }
      ],
    });
  });

  it('complex nested jsx', () => {
    const code = `
      <>
        <title>My app</title>
        <header data-i18n>
          <h1>My app</h1>
          <p>This is a demo app</p>
        </header>
        <p data-i18n>
          Some other text
          <span>© 2030</span>
        </p>
        <main>
          <p>
            Click
            <a href="/more">
              here
            </a>
            to learn more
          </p>
          <span 
            title="Some text for the span" 
          >
            Some text
          </span>
        </main>
      </>
    `;

    const ast = generateAstFromCode(code);
    const scope = extractI18n(ast);
    const scopeObj = scope?.toJSON();

    expect(scopeObj).toMatchSnapshot();
  })
});


/**
 * TODO:
 * 
 * Start with writing the entrypoint code in a declarative way.
 * Finish with adding unit tests for the LOM tree generation.
 * 
 * - Extract I18n Object Model from JSX AST (Scopes and Fragments) (Generate IDs for each scope and fragment)
 * - Loop through IOM scopes and inject React components for each fragment
 * - Generate a JSON file with the IOM tree
 * - Generate a JSON file with default values for each scope
 */