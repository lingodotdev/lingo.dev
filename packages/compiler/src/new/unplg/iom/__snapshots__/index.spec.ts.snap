// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`iom > complex nested jsx 1`] = `
{
  "explicit": false,
  "fragments": [],
  "hint": "",
  "id": "",
  "label": "",
  "role": "scope",
  "scopes": [
    {
      "explicit": false,
      "fragments": [
        {
          "id": "0",
          "role": "fragment",
          "type": "jsx/text",
          "value": "My app",
        },
      ],
      "hint": "",
      "id": "0",
      "label": "",
      "role": "scope",
      "scopes": [],
      "type": "jsx/element",
    },
    {
      "explicit": true,
      "fragments": [],
      "hint": "",
      "id": "1",
      "label": "",
      "role": "scope",
      "scopes": [
        {
          "explicit": false,
          "fragments": [
            {
              "id": "0",
              "role": "fragment",
              "type": "jsx/text",
              "value": "My app",
            },
          ],
          "hint": "",
          "id": "1.0",
          "label": "",
          "role": "scope",
          "scopes": [],
          "type": "jsx/element",
        },
        {
          "explicit": false,
          "fragments": [
            {
              "id": "0",
              "role": "fragment",
              "type": "jsx/text",
              "value": "This is a demo app",
            },
          ],
          "hint": "",
          "id": "1.1",
          "label": "",
          "role": "scope",
          "scopes": [],
          "type": "jsx/element",
        },
      ],
      "type": "jsx/element",
    },
    {
      "explicit": false,
      "fragments": [
        {
          "id": "0",
          "role": "fragment",
          "type": "jsx/text",
          "value": "Some other text",
        },
        {
          "id": "1",
          "role": "fragment",
          "type": "jsx/text",
          "value": "© 2030",
        },
      ],
      "hint": "",
      "id": "2",
      "label": "",
      "role": "scope",
      "scopes": [],
      "type": "jsx/element",
    },
    {
      "explicit": false,
      "fragments": [
        {
          "id": "0",
          "role": "fragment",
          "type": "jsx/text",
          "value": "Click",
        },
        {
          "id": "1",
          "role": "fragment",
          "type": "jsx/text",
          "value": "here",
        },
        {
          "id": "2",
          "role": "fragment",
          "type": "jsx/text",
          "value": "to learn more",
        },
      ],
      "hint": "",
      "id": "3",
      "label": "",
      "role": "scope",
      "scopes": [],
      "type": "jsx/element",
    },
    {
      "explicit": false,
      "fragments": [
        {
          "id": "0",
          "role": "fragment",
          "type": "jsx/text",
          "value": "Some text",
        },
      ],
      "hint": "",
      "id": "4",
      "label": "",
      "role": "scope",
      "scopes": [
        {
          "explicit": false,
          "fragments": [
            {
              "id": "0",
              "role": "fragment",
              "type": "js/text",
              "value": "Some text for the span",
            },
          ],
          "hint": "",
          "id": "4.0",
          "label": "title",
          "role": "scope",
          "scopes": [],
          "type": "jsx/attribute",
        },
      ],
      "type": "jsx/element",
    },
  ],
  "type": "js/program",
}
`;

exports[`parse-i18n-scope > complex nested jsx 1`] = `
{
  "explicit": false,
  "fragments": [],
  "hint": "",
  "id": "",
  "role": "scope",
  "scopes": [
    {
      "explicit": false,
      "fragments": [
        {
          "id": "0",
          "role": "fragment",
          "type": "jsx/text",
          "value": "My app",
        },
      ],
      "hint": "",
      "id": "0",
      "role": "scope",
      "scopes": [],
      "type": "jsx/element",
    },
    {
      "explicit": true,
      "fragments": [],
      "hint": "",
      "id": "1",
      "role": "scope",
      "scopes": [
        {
          "explicit": false,
          "fragments": [
            {
              "id": "0",
              "role": "fragment",
              "type": "jsx/text",
              "value": "My app",
            },
          ],
          "hint": "",
          "id": "1.0",
          "role": "scope",
          "scopes": [],
          "type": "jsx/element",
        },
        {
          "explicit": false,
          "fragments": [
            {
              "id": "0",
              "role": "fragment",
              "type": "jsx/text",
              "value": "This is a demo app",
            },
          ],
          "hint": "",
          "id": "1.1",
          "role": "scope",
          "scopes": [],
          "type": "jsx/element",
        },
      ],
      "type": "jsx/element",
    },
    {
      "explicit": false,
      "fragments": [
        {
          "id": "0",
          "role": "fragment",
          "type": "jsx/text",
          "value": "Some other text",
        },
        {
          "id": "1",
          "role": "fragment",
          "type": "jsx/text",
          "value": "© 2030",
        },
      ],
      "hint": "",
      "id": "2",
      "role": "scope",
      "scopes": [],
      "type": "jsx/element",
    },
    {
      "explicit": false,
      "fragments": [
        {
          "id": "0",
          "role": "fragment",
          "type": "jsx/text",
          "value": "Click",
        },
        {
          "id": "1",
          "role": "fragment",
          "type": "jsx/text",
          "value": "here",
        },
        {
          "id": "2",
          "role": "fragment",
          "type": "jsx/text",
          "value": "to learn more",
        },
      ],
      "hint": "",
      "id": "3",
      "role": "scope",
      "scopes": [],
      "type": "jsx/element",
    },
    {
      "explicit": false,
      "fragments": [
        {
          "id": "0",
          "role": "fragment",
          "type": "jsx/text",
          "value": "Some text",
        },
      ],
      "hint": "",
      "id": "4",
      "role": "scope",
      "scopes": [
        {
          "explicit": false,
          "fragments": [
            {
              "id": "0",
              "role": "fragment",
              "type": "js/text",
              "value": "Some text for the span",
            },
          ],
          "hint": "",
          "id": "4.0",
          "role": "scope",
          "scopes": [],
          "type": "jsx/attribute",
        },
      ],
      "type": "jsx/element",
    },
  ],
  "type": "js/program",
}
`;
